### DISCUSSION TITLE
Relation include returns both relation object & relationId, leading to Uncheckedinput error

### DISCUSSION THREAD

**Discussion by thrant** [prisma/prisma#10698] [Q&A]
So, I've got a schema with a relation:

```
model Order {
   ...
  company              Company?      @relation(fields: [companyId], references: [id])
  companyId            String?
}
```
Let's say I want to pull one record out from the DB, modify it, and send it back.

First the query:
```
const result = await this.prisma.order.findUnique({where: { id: id }, include: { company: true } });
```
The resulting object however contains BOTH the companyID and the relation itself:
```
companyId: 'ckw53wzar0051zyt08q6rshiw',
company: {
    id: 'ckw53wzar0051zyt08q6rshiw',
    createdAt: 2020-04-17T08:17:28.000Z,
    ...
```
Then I would change any data on the Order object, run it through a Save method which converts any existing relations into Prisma's "Connect / disconnect" commands, and push  to prisma.order.update() it back, leading to this error:

```
   Invalid `this.prisma.order.update()` invocation in
    ..../order.service.ts:204:39
    ...
    → 204   order = await this.prisma.order.update({
              data: {
                id: 'ckwzirec20017gdg4y363gj3g',
                company: {
                  connect: {
                    id: 'ckw53x0mf1453zyt0gebyfjji'
                  }
                },
                companyId: 'ckw53x0mf1453zyt0gebyfjji'
                ~~~~~~~~~
              },
Unknown arg `company` in data.company for type OrderUncheckedUpdateInput.
Did you mean `companyId`? Available args:...
```
...since apparently the existence of the companyId causes the generated model to fall back to the Uncheckedinput:
```
  export type OrderUpdateArgs = {
    select?: OrderSelect | null
    include?: OrderInclude | null
    data: XOR<OrderUpdateInput, OrderUncheckedUpdateInput>
    where: OrderWhereUniqueInput
```

Is there something I'm missing, or are there any recommended patterns for this?

One ideal solution would be to omit all the <relation>Id fields from the query, since they are really part of Prisma's internal workings and not needed whenever the include property is used. I checked that there was an issue with talks about omitting certain return results (just can't find the link right now) but no resolution yet on that.

I know I could just use "select" in the queries, or manually remove all <relation>Id fields in the save method, but would really not want to go there since we already have to duplicate (triplicate? :D) all the fields to schema.prisma, NestJS model, and in the frontend's GraphQL queries..

(As a side note, this does not happen with the Frontend since the GraphQL engine knows to strip out all fields that are not explicitly asked for, thus removing the <relation>Id ones. But it is a problem in operations that happen fully inside the backend.)

**Comment by matthewmueller** [comment]
@thrant thanks for kicking off this discussion!

> run it through a Save method which converts any existing relations into Prisma's "Connect / disconnect" commands

Do you mind sharing the result of the Save method? I just want to make sure we understand what the input to update is.

**Reply by thrant** [reply]
Hi, thanks for getting back to me on this!

The input to the prisma.order.update is pretty much visible in the error message (just removed the error msg parts).
```
              data: {
                id: 'ckwzirec20017gdg4y363gj3g',
                company: {
                  connect: {
                    id: 'ckw53x0mf1453zyt0gebyfjji'
                  }
                },
                companyId: 'ckw53x0mf1453zyt0gebyfjji'
              }
```
Here's a quick pseudocode of it all:
```
// Our generic save method (really in a NestJS service class)
const saveOrder = (orderInput) => {

   // Convert to Prisma's format (i.e. change included objects to connect-params)
   const {company, ...restOfData} = orderInput;
   const dataForPrisma = {
      company: { connect: { id: company.id } };
      ...restOfData
   }

   // Save -> UncheckedInput error!
   await this.prisma.order.update({data: dataForPrisma});

}

// The code that reads & modifies a single order
const order = await this.prisma.order.findUnique({where: { id: id }, include: { company: true } });
order.someValue = toSomeOtherValue;
saveOrder(order);

```

I know there's no need to explicitly do the connect again every time, but it's less work to just pass the object back as-is instead of comparing to the earlier one and sending back only diffs, and this pattern works best with our GraphQL-powered frontend where we in practice send out the full object, and receive it back with modifications.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
